scheds/rust/scx_rustland/src/bpf/alloc.rs

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) Andrea Righi <andrea.righi@canonical.com>
+
+// This software may be used and distributed according to the terms of the
+// GNU General Public License version 2.
+
+use std::alloc::{GlobalAlloc, Layout};
+use std::cell::UnsafeCell;
+use std::sync::{Mutex, MutexGuard};
+
+/// scx_rustland: memory allocator.
+///
+/// RustLandAllocator is a very simple block-based memory allocator that relies on a pre-allocated
+/// buffer and an array to manage the status of allocated and free blocks.
+///
+/// The purpose of this allocator is to prevent the user-space scheduler from triggering page
+/// faults, which could lead to potential deadlocks under heavy system load conditions.
+///
+/// Despite its simplicity, this allocator exhibits reasonable speed and efficiency in meeting
+/// memory requests from the user-space scheduler, particularly when dealing with small, uniformly
+/// sized allocations.
+
+// Pre-allocate an area of 64MB, with a block size of 64 bytes, that should be reasonable enough to
+// handle small uniform allocations performed by the user-space scheduler without introducing too
+// much fragmentation and overhead.
+const ARENA_SIZE: usize = 64 * 1024 * 1024;
+const BLOCK_SIZE: usize = 64;
+const NUM_BLOCKS: usize = ARENA_SIZE / BLOCK_SIZE;
+
+#[repr(C, align(4096))]
+struct RustLandMemory {
+    // Pre-allocated buffer.
+    arena: UnsafeCell<[u8; ARENA_SIZE]>,
+    // Allocation map.
+    //
+    // Each slot represents the status of a memory block (true = allocated, false = free).
+    allocation_map: Mutex<[bool; NUM_BLOCKS]>,
+}
+
+unsafe impl Sync for RustLandMemory {}
+
+// Memory pool for the allocator.
+static MEMORY: RustLandMemory = RustLandMemory {
+    arena: UnsafeCell::new([0; ARENA_SIZE]),
+    allocation_map: Mutex::new([false; NUM_BLOCKS]),
+};
+
+// Main allocator class.
+pub struct RustLandAllocator;
+
+impl RustLandAllocator {
+    pub fn lock_memory(&self) {
+        unsafe {
+            // Call setrlimit to set the locked-in-memory limit to unlimited.
+            let new_rlimit = libc::rlimit {
+                rlim_cur: libc::RLIM_INFINITY,
+                rlim_max: libc::RLIM_INFINITY,
+            };
+            let res = libc::setrlimit(libc::RLIMIT_MEMLOCK, &new_rlimit);
+            if res != 0 {
+                panic!("setrlimit failed with error code: {}", res);
+            }
+
+            // Lock all memory to prevent being paged out.
+            let res = libc::mlockall(libc::MCL_CURRENT | libc::MCL_FUTURE);
+            if res != 0 {
+                panic!("mlockall failed with error code: {}", res);
+            }
+        };
+    }
+}
+
+// Override global allocator methods.
+unsafe impl GlobalAlloc for RustLandAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let align = layout.align();
+        if align > BLOCK_SIZE {
```

Comment:
>64 align should be really rare so this should be fine but larger alignments can be supported relatively easily by updating the scan to find the first block to skip to the next alignment position.

Reply:
True. I completely ignored align because right now the scheduler doesn't request any alignment at all, but I can add that considering that it's a fairly easy change.